curvefs/src/client/fuse_client.cpp

```diff
@@ -946,68 +948,154 @@ CURVEFS_ERROR FuseClient::CalOneLayerSumInfo(Inode *inode) {
-                    // record hardlink
-                    if (it.type() != FsFileType::TYPE_DIRECTORY &&
-                        it.nlink() > 1) {
-                        if (hardLinkMap.count(it.inodeid())) {
-                            hardLinkMap[it.inodeid()] += it.length();
+                    inodeIds.emplace(it.inodeid());
+                }
+            }
+            inflightNum->fetch_sub(1, std::memory_order_seq_cst);
+            if (!inodeIds.empty()) {
+                auto tret = inodeManager_->BatchGetInodeAttr(&inodeIds, &attrs);
+                if (tret == CURVEFS_ERROR::OK) {
+                    std::lock_guard<std::mutex> guard(*valueMutex);
+                    for (const auto &it : attrs) {
+                        if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                            (*rsubdirs)++;
                         } else {
-                            hardLinkMap.emplace(it.inodeid(), 0);
+                            (*rfiles)++;
+                        }
+                        (*rentries)++;
+                        (*rfbytes) += it.length();
+                        // record hardlink
+                        if (it.type() != FsFileType::TYPE_DIRECTORY &&
+                            it.nlink() > 1) {
+                            std::lock_guard<std::mutex> guard(*mapMutex);
+                            auto iter = hardLinkMap->find(it.inodeid());
+                            if (iter != hardLinkMap->end()) {
+                                iter->second += it.length();
+                            } else {
+                                hardLinkMap->emplace(it.inodeid(), 0);
+                            }
                         }
                     }
+                } else {
+                    ret->exchange(false);
+                    break;
                 }
-                inodeIds.clear();
-                attrs.clear();
-            } else {
-                return ret;
             }
         }
     }
+}
+
+CURVEFS_ERROR FuseClient::CalAllLayerSumInfo(Inode *inode) {
+    std::stack<uint64_t> iStack;
+    std::mutex stackMutex;
+
+    // record hard link, <inodeId, need2minus>
+    std::unordered_map<uint64_t, uint64_t> hardLinkMap;
+    std::mutex mapMutex;
+
+    uint64_t rfiles = 0;
+    uint64_t rsubdirs = 0;
+    uint64_t rentries = 0;
+    uint64_t rfbytes = 0;
+    std::mutex valueMutex;
+
+    auto ino = inode->inodeid();
+    iStack.emplace(ino);
+    std::vector<Thread> threadpool;
+    std::atomic<uint32_t> inflightNum(0);
+    std::atomic<bool> ret(true);
+
+    auto threadNum = option_.listDentryThreads;
+    for (auto i = threadNum; i > 0; i--) {
+        threadpool.emplace_back(Thread(
+                        &FuseClient::ConcurrentGetInodeAttr,
+                        this, &iStack, &stackMutex, &hardLinkMap, &mapMutex,
+                        false, &rfiles, &rsubdirs, &rentries, &rfbytes,
+                        &valueMutex, &inflightNum, &ret));
```

Comment:
it's better to pack all the parameters into a struct.

and another suggestion is, the logic here is relatively independent, you can extract the logic into a small class, it will be more readable and maintainable

Reply:
fixed, use the class 'XattrManager' to deal with all the request about xattr.